src/cr-image/map.cr

```diff
@@ -0,0 +1,559 @@
+module CrImage
+  # An interface for 2 dimensional data. Used for various masking and correlation methods.
+  module Map(T)
+    abstract def width : Int32
+    abstract def height : Int32
+    abstract def size : Int32
+    abstract def shape : Tuple(Int32, Int32)
+    abstract def [](index : Int32) : T
+    abstract def [](x : Int32, y : Int32) : T
+    abstract def []?(index : Int32) : T?
+    abstract def []?(x : Int32, y : Int32) : T?
+
+    private def resolve_to_start_and_count(range, size) : Tuple(Int32, Int32)
+      start, count = Indexable.range_to_index_and_count(range, size) || raise IndexError.new("Unable to resolve range #{range} for mask dimension of #{size}")
+      raise IndexError.new("Range #{range} exceeds bounds of #{size}") if (start + count) > size
+      {start, count}
+    end
+  end
+
+  # :nodoc:
+  module MapImpl(T)
+    include Map(T)
+
+    macro included
+      {% verbatim do %}
+        macro method_missing(call)
+          def {{call.name.id}}{% if call.args.size > 0 %}({{call.args.join(", ").id}}){% end %} : self
+            self.class.new(width, @raw.map(&.{{call.name.id}}{% if call.args.size > 0 %}({{call.args.join(", ").id}}){% end %}))
+          end
+        end
```

Comment:
Not sure if I should keep this macro either. It's convenient when you want to transform every element when the underlying `T` type has a method, but if that method doesn't exist for `T`, it creates a difficult to figure out error message (`<method>` doesn't exist for MapImpl _or_ type `T`).

Being more explicit in which methods to forward to `T` is probably the better long term approach.